File: gherkin/java/src/main/java/io/cucumber/gherkin/pickles/PickleCompiler.java
package io.cucumber.gherkin.pickles;

import io.cucumber.messages.IdGenerator;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static io.cucumber.messages.Messages.DataTable;
import static io.cucumber.messages.Messages.DocString;
import static io.cucumber.messages.Messages.Examples;
import static io.cucumber.messages.Messages.Feature;
import static io.cucumber.messages.Messages.FeatureChild;
import static io.cucumber.messages.Messages.GherkinDocument;
import static io.cucumber.messages.Messages.Pickle;
import static io.cucumber.messages.Messages.PickleDocString;
import static io.cucumber.messages.Messages.PickleStep;
import static io.cucumber.messages.Messages.PickleStepArgument;
import static io.cucumber.messages.Messages.PickleTable;
import static io.cucumber.messages.Messages.PickleTableCell;
import static io.cucumber.messages.Messages.PickleTableRow;
import static io.cucumber.messages.Messages.PickleTag;
import static io.cucumber.messages.Messages.Rule;
import static io.cucumber.messages.Messages.RuleChild;
import static io.cucumber.messages.Messages.Scenario;
import static io.cucumber.messages.Messages.Step;
import static io.cucumber.messages.Messages.TableCell;
import static io.cucumber.messages.Messages.TableRow;
import static io.cucumber.messages.Messages.Tag;
import static java.util.Arrays.asList;
import static java.util.Collections.singletonList;
import static java.util.Collections.unmodifiableList;

public class PickleCompiler {

    private final IdGenerator idGenerator;

    public PickleCompiler(IdGenerator idGenerator) {
        this.idGenerator = idGenerator;
    }

    public List<Pickle> compile(GherkinDocument gherkinDocument, String uri) {
        List<Pickle> pickles = new ArrayList<>();
        if (!gherkinDocument.getFeature().isPresent()) {
            return pickles;
        }
        Feature feature = gherkinDocument.getFeature().get();

        String language = feature.getLanguage();

        compileFeature(pickles, feature, language, uri);
        return pickles;
    }

    private void compileFeature(List<Pickle> pickles, Feature feature, String language, String uri) {
        List<Tag> tags = feature.getTags();
        List<Step> featureBackgroundSteps = new ArrayList<>();
        for (FeatureChild featureChild : feature.getChildren()) {
            if (featureChild.getBackground().isPresent()) {
                featureBackgroundSteps.addAll(featureChild.getBackground().get().getSteps());
            } else if (featureChild.getRule().isPresent()) {
                compileRule(pickles, featureChild.getRule().get(), tags, featureBackgroundSteps, language, uri);
            } else if (featureChild.getScenario().isPresent()) {
                Scenario scenario = featureChild.getScenario().get();
                if (scenario.getExamples().isEmpty()) {
                    compileScenario(pickles, scenario, tags, featureBackgroundSteps, language, uri);
                } else {
                    compileScenarioOutline(pickles, scenario, tags, featureBackgroundSteps, language, uri);
                }
            }
        }
    }

    private void compileRule(List<Pickle> pickles, Rule rule, List<Tag> parentTags, List<Step> featureBackgroundSteps, String language, String uri) {
        List<Step> ruleBackgroundSteps = new ArrayList<>(featureBackgroundSteps);

        List<Tag> ruleTags = new ArrayList<>();
        ruleTags.addAll(parentTags);
        ruleTags.addAll(rule.getTags());

        for (RuleChild ruleChild : rule.getChildren()) {
            if (ruleChild.getBackground().isPresent()) {
                ruleBackgroundSteps.addAll(ruleChild.getBackground().get().getSteps());
            } else if (ruleChild.getScenario().isPresent()) {
                Scenario scenario = ruleChild.getScenario().get();
                if (scenario.getExamples().isEmpty()) {
                    compileScenario(pickles, scenario, ruleTags, ruleBackgroundSteps, language, uri);
                } else {
                    compileScenarioOutline(pickles, scenario, ruleTags, ruleBackgroundSteps, language, uri);
                }
            }
        }
    }

    private void compileScenario(List<Pickle> pickles, Scenario scenario, List<Tag> parentTags, List<Step> backgroundSteps, String language, String uri) {
        List<PickleStep> steps = new ArrayList<>();
        if (!scenario.getSteps().isEmpty())
            steps.addAll(pickleSteps(backgroundSteps));

        steps.addAll(pickleSteps(scenario.getSteps()));

        List<Tag> scenarioTags = new ArrayList<>();
        scenarioTags.addAll(parentTags);
        scenarioTags.addAll(scenario.getTags());

        List<String> sourceIds = singletonList(scenario.getId());

        Pickle pickle = new Pickle(
                idGenerator.newId(),
                uri,
                scenario.getName(),
                language,
                steps,
                pickleTags(scenarioTags),
                sourceIds
        );
        pickles.add(pickle);
    }

    private void compileScenarioOutline(List<Pickle> pickles, Scenario scenario, List<Tag> featureTags, List<Step> backgroundSteps, String language, String uri) {
        for (final Examples examples : scenario.getExamples()) {
            if (!examples.getTableHeader().isPresent()) continue;
            List<TableCell> variableCells = examples.getTableHeader().get().getCells();
            for (final TableRow valuesRow : examples.getTableBody()) {
                List<TableCell> valueCells = valuesRow.getCells();

                List<PickleStep> steps = new ArrayList<>();

                if (!scenario.getSteps().isEmpty())
                    steps.addAll(pickleSteps(backgroundSteps));


                List<Tag> tags = new ArrayList<>();
                tags.addAll(featureTags);
                tags.addAll(scenario.getTags());
                tags.addAll(examples.getTags());

                for (Step scenarioOutlineStep : scenario.getSteps()) {
                    PickleStep pickleStep = pickleStep(scenarioOutlineStep, variableCells, valuesRow);

                    steps.add(pickleStep);
                }

                List<String> sourceIds = asList(scenario.getId(), valuesRow.getId());
                Pickle pickle = new Pickle(
                        idGenerator.newId(),
                        uri,
                        interpolate(scenario.getName(), variableCells, valueCells),
                        language,
                        steps,
                        pickleTags(tags),
                        sourceIds
                );

                pickles.add(pickle);
            }
        }
    }

    private PickleTable pickleDataTable(DataTable dataTable, List<TableCell> variableCells, List<TableCell> valueCells) {
        List<TableRow> rows = dataTable.getRows();
        List<PickleTableRow> newRows = new ArrayList<>(rows.size());
        for (TableRow row : rows) {
            List<TableCell> cells = row.getCells();
            List<PickleTableCell> newCells = new ArrayList<>();
            for (TableCell cell : cells) {
                newCells.add(new PickleTableCell(interpolate(cell.getValue(), variableCells, valueCells)));
            }
            newRows.add(new PickleTableRow(newCells));
        }
        return new PickleTable(newRows);
    }

    private PickleDocString pickleDocString(DocString docString, List<TableCell> variableCells, List<TableCell> valueCells) {
        return new PickleDocString(
                docString.getMediaType().isPresent() ? interpolate(docString.getMediaType().get(), variableCells, valueCells) : null,
                interpolate(docString.getContent(), variableCells, valueCells)
        );
    }

    private PickleStep pickleStep(Step step, List<TableCell> variableCells, TableRow valuesRow) {
        List<TableCell> valueCells = valuesRow == null ? Collections.emptyList() : valuesRow.getCells();
        String stepText = interpolate(step.getText(), variableCells, valueCells);

        PickleStepArgument argument = null;
        if (step.getDataTable().isPresent()) {
            argument = new PickleStepArgument(null, pickleDataTable(step.getDataTable().get(), variableCells, valueCells));
        }

        if (step.getDocString().isPresent()) {
            argument = new PickleStepArgument(pickleDocString(step.getDocString().get(), variableCells, valueCells), null);
        }


        List<String> astNodeIds;
        if (valuesRow != null) {
            astNodeIds = Stream.of(singletonList(step.getId()), singletonList(valuesRow.getId()))
                    .flatMap(Collection::stream)
                    .collect(Collectors.toList());

        } else {
            astNodeIds = singletonList(step.getId());
        }

        return new PickleStep(
                argument,
                astNodeIds,
                idGenerator.newId(),
                stepText
        );
    }

    private List<PickleStep> pickleSteps(List<Step> steps) {
        List<PickleStep> result = new ArrayList<>();
        for (Step step : steps) {
            result.add(pickleStep(step));
        }
        return unmodifiableList(result);
    }

    private PickleStep pickleStep(Step step) {
        return pickleStep(step, Collections.emptyList(), null);
    }

    private String interpolate(String name, List<TableCell> variableCells, List<TableCell> valueCells) {
        int col = 0;
        for (TableCell variableCell : variableCells) {
            TableCell valueCell = valueCells.get(col++);
            String header = variableCell.getValue();
            String value = valueCell.getValue();
            name = name.replace("<" + header + ">", value);
        }
        return name;
    }

    private List<PickleTag> pickleTags(List<Tag> tags) {
        List<PickleTag> result = new ArrayList<>();
        for (Tag tag : tags) {
            result.add(pickleTag(tag));
        }
        return result;
    }

    private PickleTag pickleTag(Tag tag) {
        return new PickleTag(tag.getName(), tag.getId());
    }

}

File: gherkin/php/src/PickleCompiler.php
<?php

declare(strict_types=1);

namespace Cucumber\Gherkin;

use Cucumber\Messages\DataTable;
use Cucumber\Messages\DocString;
use Cucumber\Messages\Feature;
use Cucumber\Messages\GherkinDocument;
use Cucumber\Messages\Id\IdGenerator;
use Cucumber\Messages\Pickle;
use Cucumber\Messages\PickleDocString;
use Cucumber\Messages\PickleStep;
use Cucumber\Messages\PickleStepArgument;
use Cucumber\Messages\PickleTable;
use Cucumber\Messages\PickleTableCell;
use Cucumber\Messages\PickleTableRow;
use Cucumber\Messages\PickleTag;
use Cucumber\Messages\Rule;
use Cucumber\Messages\Scenario;
use Cucumber\Messages\Step;
use Cucumber\Messages\TableCell;
use Cucumber\Messages\TableRow;
use Cucumber\Messages\Tag;
use Generator;

final class PickleCompiler
{
    /** @var array<string, PickleStep\Type|null> */
    private array $pickleStepTypeFromKeyword = [];

    private Step\KeywordType $lastKeywordType = Step\KeywordType::UNKNOWN;

    public function __construct(
        private readonly IdGenerator $idGenerator,
    ) {
        $this->pickleStepTypeFromKeyword[Step\KeywordType::UNKNOWN->name] = PickleStep\Type::UNKNOWN;
        $this->pickleStepTypeFromKeyword[Step\KeywordType::CONTEXT->name] = PickleStep\Type::CONTEXT;
        $this->pickleStepTypeFromKeyword[Step\KeywordType::ACTION->name] = PickleStep\Type::ACTION;
        $this->pickleStepTypeFromKeyword[Step\KeywordType::OUTCOME->name] = PickleStep\Type::OUTCOME;
        $this->pickleStepTypeFromKeyword[Step\KeywordType::CONJUNCTION->name] = null;
    }

    /**
     * @return Generator<Pickle>
     */
    public function compile(GherkinDocument $gherkinDocument, string $uri): Generator
    {
        if (!$gherkinDocument->feature) {
            return;
        }

        $feature = $gherkinDocument->feature;
        $language = $feature->language;

        yield from $this->compileFeature($feature, $language, $uri);
    }

    /**
     * @return Generator<Pickle>
     */
    private function compileFeature(Feature $feature, string $language, string $uri): Generator
    {
        $tags = $feature->tags;
        $featureBackgroundSteps = [];

        foreach ($feature->children as $featureChild) {
            if ($featureChild->background) {
                $featureBackgroundSteps = [...$featureBackgroundSteps, ...$featureChild->background->steps];
            } elseif ($featureChild->rule) {
                yield from $this->compileRule($featureChild->rule, $tags, $featureBackgroundSteps, $language, $uri);
            } elseif ($featureChild->scenario) {
                if (!$featureChild->scenario->examples) {
                    yield from $this->compileScenario($featureChild->scenario, $tags, $featureBackgroundSteps, $language, $uri);
                } else {
                    yield from $this->compileScenarioOutline($featureChild->scenario, $tags, $featureBackgroundSteps, $language, $uri);
                }
            }
        }
    }

    /**
     * @param list<Tag> $parentTags
     * @param list<Step> $featureBackgroundSteps
     *
     * @return Generator<Pickle>
     */
    private function compileRule(Rule $rule, array $parentTags, array $featureBackgroundSteps, string $language, string $uri): Generator
    {
        $ruleBackgroundSteps = $featureBackgroundSteps;
        $ruleTags = [...$parentTags, ...$rule->tags];

        foreach ($rule->children as $ruleChild) {
            if ($ruleChild->background) {
                $ruleBackgroundSteps = [...$ruleBackgroundSteps, ...$ruleChild->background->steps];
            } elseif ($ruleChild->scenario && $ruleChild->scenario->examples) {
                yield from $this->compileScenarioOutline($ruleChild->scenario, $ruleTags, $ruleBackgroundSteps, $language, $uri);
            } elseif ($ruleChild->scenario) {
                yield from $this->compileScenario($ruleChild->scenario, $ruleTags, $ruleBackgroundSteps, $language, $uri);
            }
        }
    }

    /**
     * @param list<Tag> $parentTags
     * @param list<Step> $backgroundSteps
     *
     * @return Generator<Pickle>
     */
    private function compileScenario(Scenario $scenario, array $parentTags, array $backgroundSteps, string $language, string $uri): Generator
    {
        $steps = [
            ...($scenario->steps ? $this->pickleSteps($backgroundSteps) : []),
            ...array_map(fn ($s) => $this->pickleStep($s), $scenario->steps),
        ];

        $tags = [...$parentTags, ...$scenario->tags];
        $this->lastKeywordType = Step\KeywordType::UNKNOWN;

        yield new Pickle(
            id: $this->idGenerator->newId(),
            uri: $uri,
            name: $scenario->name,
            language: $language,
            steps: $steps,
            tags: $this->pickleTags($tags),
            astNodeIds: [$scenario->id],
        );
    }

    /**
     * @param list<Tag> $featureTags
     * @param list<Step> $backgroundSteps
     *
     * @return Generator<Pickle>
     */
    private function compileScenarioOutline(Scenario $scenario, array $featureTags, array $backgroundSteps, string $language, string $uri): Generator
    {
        foreach ($scenario->examples as $examples) {
            if (!$examples->tableHeader) {
                continue;
            }

            $variableCells = $examples->tableHeader->cells;
            foreach ($examples->tableBody as $valuesRow) {
                $valueCells = $valuesRow->cells;

                $steps = [
                    ...($scenario->steps ? $this->pickleSteps($backgroundSteps) : []),
                    ...array_map(fn ($s) => $this->pickleStep($s, $variableCells, $valuesRow), $scenario->steps),
                ];

                $tags = [...$featureTags, ...$scenario->tags, ...$examples->tags];

                $sourceIds = [$scenario->id, $valuesRow->id];

                yield new Pickle(
                    id: $this->idGenerator->newId(),
                    uri: $uri,
                    name: $this->interpolate($scenario->name, $variableCells, $valueCells),
                    language: $language,
                    steps: $steps,
                    tags: $this->pickleTags($tags),
                    astNodeIds: $sourceIds,
                );
            }
        }
    }

    /**
     * @param list<Step> $steps
     * @return list<PickleStep>
     */
    private function pickleSteps(array $steps): array
    {
        return array_map(fn ($s) => $this->pickleStep($s), $steps);
    }

    /**
     * @param list<TableCell> $variableCells
     */
    private function pickleStep(Step $step, array $variableCells = [], ?TableRow $valuesRow = null): PickleStep
    {
        $valueCells = $valuesRow?->cells ?? [];
        $stepText = $this->interpolate($step->text, $variableCells, $valueCells);

        $astNodeIds = $valuesRow ? [$step->id, $valuesRow->id] : [$step->id];

        if ($step->dataTable) {
            $argument = new PickleStepArgument(dataTable: $this->pickleDataTable($step->dataTable, $variableCells, $valueCells));
        } elseif ($step->docString) {
            $argument = new PickleStepArgument(docString: $this->pickleDocString($step->docString, $variableCells, $valueCells));
        } else {
            $argument = null;
        }

        $this->lastKeywordType =
            $step->keywordType === Step\KeywordType::CONJUNCTION
            ? $this->lastKeywordType
            : ($step->keywordType ?? Step\KeywordType::UNKNOWN)
        ;

        return new PickleStep(
            argument: $argument,
            astNodeIds: $astNodeIds,
            id: $this->idGenerator->newId(),
            type: $this->pickleStepTypeFromKeyword[$this->lastKeywordType->name],
            text: $stepText,
        );
    }

    /**
     * @param list<TableCell> $variableCells
     * @param list<TableCell> $valueCells
     */
    private function interpolate(string $name, array $variableCells, array $valueCells): string
    {
        $variables = array_map(fn ($c) => '<'.$c->value.'>', $variableCells);
        $values = array_map(fn ($c) => $c->value, $valueCells);

        $replacements = array_combine($variables, $values);

        return StringUtils::replace($name, $replacements);
    }

    /**
     * @param list<TableCell> $variableCells
     * @param list<TableCell> $valueCells
     */
    private function pickleDataTable(DataTable $dataTable, array $variableCells, array $valueCells): PickleTable
    {
        return new PickleTable(
            rows: array_map(
                fn ($r) => new PickleTableRow(
                    cells: array_map(
                        fn ($c) => new PickleTableCell(
                            $this->interpolate($c->value, $variableCells, $valueCells),
                        ),
                        $r->cells,
                    ),
                ),
                $dataTable->rows,
            ),
        );
    }

    /**
     * @param list<TableCell> $variableCells
     * @param list<TableCell> $valueCells
     */
    private function pickleDocString(DocString $docstring, array $variableCells, array $valueCells): PickleDocString
    {
        return new PickleDocString(
            mediaType: $docstring->mediaType ? $this->interpolate($docstring->mediaType, $variableCells, $valueCells) : null,
            content: $this->interpolate($docstring->content, $variableCells, $valueCells),
        );
    }

    /**
     * @param list<Tag> $tags
     * @return list<PickleTag>
     */
    private function pickleTags(array $tags): array
    {
        return array_map(fn ($t) => $this->pickleTag($t), $tags);
    }

    private function pickleTag(Tag $tag): PickleTag
    {
        return new PickleTag($tag->name, $tag->id);
    }
}
